### PR TITLE
cses-1715 and combo module stuff

### PR DESCRIPTION
**Pull request by envyaims**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making them.

### CODE REVIEW

File: solutions/cses-1715.mdx [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+const ll MOD = 1e9 + 7;
+const int MAXN = 1e6;
+ 
+array<ll, MAXN + 1> fact;  // fact[i] = i! % MOD
+array<ll, MAXN + 1> inv;  // inv[i] = modular inverse of fact[i];
+ 
+// calculate a ^ b (mod m) in log b time
+ll modpow(ll a, ll b, ll m){
+	ll prod = 1;
+	while(b > 0){
```
Comment (SansPapyrus683) [inline solutions/cses-1715.mdx:35]: spaces for control flow

File: solutions/cses-1715.mdx [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	string s;
+	cin >> s;
+	
+	vector<int> character_count(26);
+	for(char i: s){
+		character_count[i - 'a']++;
+	}
+	
+	ll total = fact[s.length()];
+	for(int i: character_count){
```
Comment (SansPapyrus683) [inline solutions/cses-1715.mdx:62]:
```suggestion
	for (int i : character_count) {
```

File: solutions/cses-1715.mdx [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+## Explanation
+
+This strategy is known as [The Multinomial Theorem](https://en.wikipedia.org/wiki/Multinomial_theorem).
+
+We are essentially counting how many distinct strings we can make by permuting the letters of the string. Note that if there are repeating characters in the string, permuting those characters in place counts as the same string since letters are indisguinshable. 
+
+For example, consider the string $aab$, and supposed we denote the $a$'s that sppear in the string in order of which they appear, so the original string becomes $a_1a_2b$. One possible permutation is $a_2a_1b$. When we remove these subscripts, these two strings become the same thing. 
+
+Therefore, we will need to divide those possibilities off. The total number of permutations is $n!$. The number of ways a letter can permute is also just the factorial of the number of times it appears in the string.
```
Comment (SansPapyrus683) [inline solutions/cses-1715.mdx:16]: > Therefore, we will need to divide those possibilities off. 

how about "we'll need to divide by [what] to account for this"

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
honestly pretty good

File: solutions/cses-1715.mdx [hunk 4]
```diff
@@ -13,7 +13,7 @@ We are essentially counting how many distinct strings we can make by permuting t
 
 For example, consider the string $aab$, and supposed we denote the $a$'s that sppear in the string in order of which they appear, so the original string becomes $a_1a_2b$. One possible permutation is $a_2a_1b$. When we remove these subscripts, these two strings become the same thing. 
 
-Therefore, we will need to divide those possibilities off. The total number of permutations is $n!$. The number of ways a letter can permute is also just the factorial of the number of times it appears in the string. 
+Therefore, we will need to divide off the permutations of each repeating letter for this. The total number of permutations is $n!$. The number of ways a letter can permute is also just the factorial of the number of times it appears in the string.
```
Comment (SansPapyrus683) [inline solutions/cses-1715.mdx:16]: > the permutations of each repeating letter for this

what

Comment (envyaims) [reply solutions/cses-1715.mdx:16]: how to english 

:divide off the number of arrangements for each individual letter"?

Comment (SansPapyrus683) [reply solutions/cses-1715.mdx:16]: what does permutations of a letter mean?
if it's a single character, there's just 1 permutation

Comment (envyaims) [reply solutions/cses-1715.mdx:16]: each type of letter in the alphabet

Comment (SansPapyrus683) [reply solutions/cses-1715.mdx:16]: so say if there were 2 `a`s, we'd divide by $2!$?

Comment (envyaims) [reply solutions/cses-1715.mdx:16]: yeah

Comment (envyaims) [reply solutions/cses-1715.mdx]: better?

**Review by SansPapyrus683 — APPROVED** [review]

File: solutions/cses-1715.mdx [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+array<ll, MAXN + 1> inv;  // inv[i] = modular inverse of fact[i];
+ 
+// calculate a ^ b (mod m) in log b time
+ll modpow(ll a, ll b, ll m) {
+	ll prod = 1;
+	while (b > 0) {
+		if (b % 2 == 1) {
+			prod = (prod * a) % m;
+		}
+		b /= 2;
```
Comment (Aadit-Ambadkar) [inline solutions/cses-1715.mdx:39]:
```suggestion
		b >>= 1;
```

File: solutions/cses-1715.mdx [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+const int MAXN = 1e6;
+ 
+array<ll, MAXN + 1> fact;  // fact[i] = i! % MOD
+array<ll, MAXN + 1> inv;  // inv[i] = modular inverse of fact[i];
+ 
+// calculate a ^ b (mod m) in log b time
+ll modpow(ll a, ll b, ll m) {
+	ll prod = 1;
+	while (b > 0) {
+		if (b % 2 == 1) {
```
Comment (Aadit-Ambadkar) [inline solutions/cses-1715.mdx:36]:
```suggestion
		if (b & 1) {
```

File: solutions/cses-1715.mdx [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+			prod = (prod * a) % m;
+		}
+		b /= 2;
+		a = (a * a) % m;
+	}
+	return prod;
+}
+ 
+int main() {
+	// precalculate fact[i] and inv[i]
```
Comment (Aadit-Ambadkar) [inline solutions/cses-1715.mdx:46]:
```suggestion
	// precalcompute fact[i] and inv[i]
```

File: solutions/cses-1715.mdx [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+	for (int i = 2; i <= MAXN; i++) {
+		fact[i] = fact[i-1] * i % MOD;
+		inv[i] = modpow(fact[i], MOD - 2, MOD);
+	}
+	
+	string s;
+	cin >> s;
+	
+	vector<int> character_count(26);
+	for (char i: s) {
```
Comment (Aadit-Ambadkar) [inline solutions/cses-1715.mdx:57]:
```suggestion
	for (char i : s) {
```

File: solutions/cses-1715.mdx [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+		a = (a * a) % m;
+	}
+	return prod;
+}
+ 
+int main() {
+	// precalculate fact[i] and inv[i]
+	fact[0] = inv[0] = fact[1] = inv[1] = 1;
+	for (int i = 2; i <= MAXN; i++) {
+		fact[i] = fact[i-1] * i % MOD;
```
Comment (Aadit-Ambadkar) [inline solutions/cses-1715.mdx:49]:
```suggestion
		fact[i] = fact[i - 1] * i % MOD;
```

**Review by Aadit-Ambadkar — CHANGES_REQUESTED** [review]

File: solutions/cses-1715.mdx (hunk 5 above)
Comment (SansPapyrus683) [reply solutions/cses-1715.mdx:39]: hell naw man

File: solutions/cses-1715.mdx (hunk 6 above)
Comment (SansPapyrus683) [reply solutions/cses-1715.mdx:36]: aadit you're actually trollling

File: solutions/cses-1715.mdx (hunk 7 above)
Comment (SansPapyrus683) [reply solutions/cses-1715.mdx:46]: > precalcompute

?

Comment (envyaims) [reply solutions/cses-1715.mdx:46]: 🤔

**Comment by SansPapyrus683** [comment]
no i just haven't approved yet

File: solutions/cses-1715.mdx (hunk 7 above)
Comment (Aadit-Ambadkar) [reply solutions/cses-1715.mdx:46]: 🤔

Comment (Aadit-Ambadkar) [reply solutions/cses-1715.mdx:46]: sorry about that @envyaims

**Review by jessechoe10 — APPROVED** [review]

**Comment by SansPapyrus683** [comment]
image
...

**Comment by envyaims** [comment]
> image
> 
> ...

what am i supposed to do about it

**Comment by envyaims** [comment]
@SansPapyrus683 fixed

**Pull request merged**